beacon-chain/blockchain/fork_choice.go

```diff
@@ -0,0 +1,105 @@
+package blockchain
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/prysmaticlabs/prysm/beacon-chain/db"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	bytesutil "github.com/prysmaticlabs/prysm/shared/bytes"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+)
+
+// LMDGhost applies the Latest Message Driven, Greediest Heaviest Observed Sub-Tree
+// fork-choice rule defined in the Ethereum Serenity specification for the beacon chain.
+func LMDGhost(
+	block *pb.BeaconBlock,
+	voteTargets map[[32]byte]*pb.BeaconBlock,
+	observedBlocks []*pb.BeaconBlock,
+	beaconDB *db.BeaconDB,
+) (*pb.BeaconBlock, error) {
+	head := block
+	for {
+		children, err := BlockChildren(head, observedBlocks)
+		if err != nil {
+			return nil, fmt.Errorf("could not fetch block children: %v", err)
+		}
+		if len(children) == 0 {
+			return head, nil
+		}
+		maxChild := children[0]
+		maxChildVotes, err := VoteCount(maxChild, voteTargets, beaconDB)
+		if err != nil {
+			return nil, fmt.Errorf("unable to determine vote count for block: %v", err)
+		}
+		for i := 0; i < len(children); i++ {
+			candidateChildVotes, err := VoteCount(children[i], voteTargets, beaconDB)
+			if err != nil {
+				return nil, fmt.Errorf("unable to determine vote count for block: %v", err)
+			}
+			if candidateChildVotes > maxChildVotes {
+				maxChild = children[i]
+			}
+		}
+		head = maxChild
+	}
+}
+
+// VoteCount determines the number of votes on a beacon block by counting the number
+// of target blocks that have such beacon block as a common ancestor.
+func VoteCount(block *pb.BeaconBlock, targets map[[32]byte]*pb.BeaconBlock, beaconDB *db.BeaconDB) (int, error) {
+	votes := 0
+	for k := range targets {
+		ancestor, err := BlockAncestor(targets[k], block.Slot, beaconDB)
+		if err != nil {
+			return 0, err
+		}
+		ancestorEnc, err := proto.Marshal(ancestor)
+		if err != nil {
+			return 0, err
+		}
+		ancestorHash := hashutil.Hash(ancestorEnc)
+		blockEnc, err := proto.Marshal(block)
+		if err != nil {
+			return 0, err
+		}
+		blockHash := hashutil.Hash(blockEnc)
+		if blockHash == ancestorHash {
+			votes++
+		}
+	}
+	return votes, nil
+}
+
+// BlockAncestor obtains the ancestor at of a block at a certain slot.
+func BlockAncestor(block *pb.BeaconBlock, slot uint64, beaconDB *db.BeaconDB) (*pb.BeaconBlock, error) {
```

Comment:
A more appropriate place for `BlockAncestor ` is under `block.go`. `BlockAncestor` is more of a block-specific operation, not fork choice

Reply:
Ancestor is very coupled with the vote count and currently depends on the DB. I don't want to introduce db methods into core, so for now we can just move children. Also, the ancestor function is the prime candidate to be changed with future optimizations.